rest/method_head.go

```diff
@@ -0,0 +1,54 @@
+package rest
+
+import (
+	"context"
+	"net/http"
+
+	"github.com/rs/rest-layer/resource"
+)
+
+// listHead handles HEAD resquests on a resource URL. Returns no payload
+func listHead(ctx context.Context, r *http.Request, route *RouteMatch) (status int, headers http.Header, body interface{}) {
+	forceTotal := false
+	rsc := route.Resource()
+	switch rsc.Conf().ForceTotal {
+	case resource.TotalOptIn:
+		forceTotal = route.Params.Get("total") == "1"
+	case resource.TotalAlways:
+		forceTotal = true
+	case resource.TotalDenied:
+		if route.Params.Get("total") == "1" {
+			return 422, nil, &Error{422, "Cannot use `total' parameter: denied by configuration", nil}
+		}
+	}
+	q, e := route.Query()
+	if e != nil {
+		return e.Code, nil, e
+	}
+	list := &resource.ItemList{
+		Total: -1,
+		Items: []*resource.Item{},
+	}
+	var err error
+	if forceTotal {
+		list.Total, err = rsc.Count(ctx, q)
+		// If Storer doesn't implement Counter interface,
+		// fallback to listGet implementation
+		if err == resource.ErrNotImplemented {
+			return listGet(ctx, r, route)
+		}
+	}
+	if err != nil {
+		e = NewError(err)
+		return e.Code, nil, e
+	}
+	if win := q.Window; win != nil {
+		if win.Offset > 0 {
```

Comment:
If offset is off bounds, shouldn’t we return a 404?

Reply:
Let me think a bit on Offsets.

Reply:
There is explicit setting for Offset&Limit here. I think this is not right. Please lets discuss.
https://github.com/rs/rest-layer/blob/0df312935ab2631d1813920adf9bb04b328dd071/rest/routing.go#L198

Reply:
@Dragomir-Ivanov, that is the _previous_ implementaton of HEAD, so I suppose you could change it _if_ you plan to include the offset/count as response headers. However, you _should_ probably call Find on the storage backend with these Window settings as a fallback when the `Counter` interface is not implemented.

Reply:
@rs, a bit of-topic for the review, but in the code that @Dragomir-Ivanov links to, I see you only handle `GET` and `HEAD`. I suppose that in order to fix #127, `page`, `limit` and `offset` (although not `fields`) should be handled in the case of `DELETE`.

Reply:
@smyrman that's a good point.

Reply:
@smyrman I think out-of-bounds offset doesn't work for current master:
```
GET http://192.168.0.103:3001/api/clients?skip=2&total=1&filter={"isBlacklisted":false}&limit=10

HTTP/1.1 200 OK
Content-Length: 660
Content-Type: application/json
Date: Wed, 18 Oct 2017 13:52:23 GMT
Request-Id: b7jlp5p40fm078fk87jg
Vary: Origin
X-Offset: 2
X-Total: 4

[
  {...},
  {...}
]
```

However this:
```
GET http://192.168.0.103:3001/api/clients?skip=100&total=1&filter={"isBlacklisted":false}&limit=10

HTTP/1.1 200 OK
Content-Length: 2
Content-Type: application/json
Date: Wed, 18 Oct 2017 13:54:20 GMT
Request-Id: b7jlq3140fm078fk87k0
Vary: Origin
X-Offset: 100
X-Total: 100

[]
```
It is buggy ( notice the value for `X-Offset`/`X-Total` ) and still returns `200 OK`.

Reply:
Okay, I think that this specialization is not needed. In the recent brainstorm about the ETags, all items will need to be extracted ( or at least their `_etag` fields ) and be combined/summed in order to support `ETag` for list resources. Sorry about the noise. 
Let me understand what out-of-bound `offset` is, so I can fix at least this.

Reply:
I think, based on a very fast Google, the best thing for out of bounds, is to continue returning `200 OK`.

Regarding the `X-Total`, is it fixed in rs/rest-layer-mongo#19 in terms of the mongo storage backend at least?

Reply:
Okay, so leaving it as it is, and ignoring Olivier's proposal for returning 404. 
@rs Are you okay with this?

Reply:
He is (see slack)

Reply:
Closing this PR, since `200 OK` is current behavior, and having optimized HEAD implementation is not possible we will need to generate `ETag` based on DB contents.